doc/source/reference/arrays.classes.rst

```diff
@@ -316,11 +316,13 @@ NumPy provides several hooks that classes can customize:
 
 .. py:method:: class.__array__([dtype])
 
-   If a class (ndarray subclass or not) having the :func:`__array__`
-   method is used as the output object of an :ref:`ufunc
-   <ufuncs-output-type>`, results will *not* be written to the object
-   returned by :func:`__array__`. This practice will return ``TypeError``.
-
+    This method is called by class object to return a numpy array.
```

Comment:
This is better than the status quo but still a little unclear IMO. 

Here's my try:

If defined on an object, should return a representation of the object as a NumPy array. The value returned by this method is used if an object implementing this interface is passed to array-coercion functions like `np.array()`.

I would then move the "Please refer..." you added to the end immediately after the above explanation.

Finally, since the content that's already in the docs is more of a usage note than an explanation for what `__array__` is, I would separate the rest into a `.. note::` section.

Reply:
Done, thanks @ngoldbaum .